### PR TITLE
[Bugfix:API] Fix DatetimeFieldOverflow error handling

### LINKED ISSUE #3331: Internal Server Error when searching with partial dates in Find a Record
## Description

Sometimes when partially entering the date in the Record Search Field, it randomly throws internal server errors.
Sometimes it throws the intended `InvalidDateFormatAPIException` but other times it `UnboundLocalError`.

Here is a recording of the behaviour:

https://github.com/mathesar-foundation/mathesar/assets/31622972/6e7885c2-9d19-473a-86e0-8c63c5e68c0f

Also, it would be good UI/UX to show the user the date is invalid. In case when the user has just put an incorrect date time format and the server returns a 400 BAD REQUEST InvalidDateFormatAPIException. The user won't know as the table still shows the loading state.
image

## Expected behavior

- Server should not throw 500 internal server error and always throw InvalidDateFormatAPIException in case of wrong date format.
- User should be conveyed if their input date is incorrect.

## Environment
 - OS: macOS
 - Browser: Chrome
 - Other info:

### PR DESCRIPTION
**Pull request by sahilsuman933**
Fixes #3331

This change incorporates an additional error from psycopg2 within the try-catch exception block in DatetimeFieldOverflow to handle out-of-range date values. Due to this, we encountered an UnboundLocalError, which shouldn't have been the case; instead, it should have been an error in the format of InvalidDateFormatAPIException.
 
**Screenshots**

Error Message:
image

After implementing this error handling case:
image

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `develop` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [X] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Anish9901** [comment]
Hey, @sahilsuman933 please add a test case for this.

File: mathesar/api/db/viewsets/records.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def list(self, request, table_pk=None):
                 status_code=status.HTTP_400_BAD_REQUEST
             )
         except DataError as e:
-            if isinstance(e.orig, InvalidDatetimeFormat):
+            if isinstance(e.orig, (InvalidDatetimeFormat, DatetimeFieldOverflow)):
```
Comment (Anish9901) [inline mathesar/api/db/viewsets/records.py:106]: Add an else to raise `e`, for the case in which we're unable to catch the specific exception.

**Review by Anish9901 — CHANGES_REQUESTED** [review]

**Comment by sahilsuman933** [comment]
There is no need for an additional test case because there is already a test case that should cover this issue. You can verify this by checking the test case `99/99/9999`. Previously, it resulted in a 500 Internal Server Error, but after my fix, it no longer throws that HTTP error.

Test Case: 
image

SRC: /mathesar/tests/api/test_record_api.py 

Before my fix:
image


After my fix:
image


// @Anish9901

File: mathesar/api/db/viewsets/records.py (hunk 1 above)
Comment (sahilsuman933) [reply mathesar/api/db/viewsets/records.py:106]: Do you mean that I should add an `except Exception as e` to handle all errors, or do you want me to create a separate `elif` statement for `DatetimeFieldOverflow`?

Comment (sahilsuman933) [reply mathesar/api/db/viewsets/records.py:106]: Is this what you want?
image

Comment (Anish9901) [reply mathesar/api/db/viewsets/records.py:106]: Your current Implementation is fine just add an else block to throw any _other_ `DataError`, I want to eliminate the senarios where we are catching the exception correctly(`DataError` in this case) but not throwing it.

**Comment by Anish9901** [comment]
> There is no need for an additional test case because there is already a test case that should cover this issue. You can verify this by checking the test case 99/99/9999. Previously, it resulted in a 500 Internal Server Error, but after my fix, it no longer throws that HTTP error.

Was this the case while running the pytest as well? As from the video in the issue description it seems like the exception was getting thown the 2nd time someone tried to add an invalid date.

**Comment by sahilsuman933** [comment]
> Was this the case while running the pytest as well? 

No, I didn't encounter any issues while running the pytest. When I manually tested it with the table data used by pytest, it ran without any problems. Surprisingly, it didn't even raise an error for the value `99/99/9999`, which should have triggered an "invalid_date_format" issue.

image

Instead of returning a count of 0, it should throw an error for an invalid date. However, this was not the case with the test database used by pytest.

I am using patents table data:
image

Now, when I tried using it with another test data, specifically the "invalid_reference_base_table.csv" table, it started throwing an "InvalidDateFormat" error.

> As from the video in the issue description it seems like the exception was getting thown the 2nd time someone tried to add an invalid date.

If you carefully observe the video, you'll notice that the date format should be in MM-DD-YYYY, and the person was attempting to enter a value greater than 12, which caused the "DatetimeFieldOverflow" error. This is because there are only 12 months in a year.

File: mathesar/api/db/viewsets/records.py (hunk 1 above)
Comment (sahilsuman933) [reply mathesar/api/db/viewsets/records.py:106]: Added these changes.

File: mathesar/api/db/viewsets/records.py [hunk 2]
```diff
@@ -103,11 +103,16 @@ def list(self, request, table_pk=None):
...
```
Comment (Anish9901) [inline mathesar/api/db/viewsets/records.py:106]: Please split these to raise `InvalidDateFormatAPIException` and `InvalidDateAPIException` respectively.

File: mathesar/api/db/viewsets/records.py [hunk 3]
```diff
@@ -103,11 +103,16 @@ def list(self, request, table_pk=None):
...
                 raise database_api_exceptions.InvalidDateFormatAPIException(
                     e,
                     status_code=status.HTTP_400_BAD_REQUEST,
                 )
+            else:
+                raise database_api_exceptions.RaiseExceptionAPIException(
```
Comment (Anish9901) [inline mathesar/api/db/viewsets/records.py:112]: `MathesarAPIException` would be more appropriate here, check the following for reference.
https://github.com/mathesar-foundation/mathesar/blob/d375575199e81205ea9cd4a7a3b928398f43ec65/mathesar/api/serializers/records.py#L62-L63

**Review by Anish9901 — CHANGES_REQUESTED** [review]
I noticed a couple more things since my last review:
- We have a separate exception class for `DatetimeFieldOverlflow` namely `InvalidDateAPIException`.
-  I was able to reproduce the issue locally using our pytest test suite, and you should be adding a test case addressing this issue, the test case mentioned in your screenshot refer to record update instead of record search you can see the difference by noticing the patch and get requests.
Look at `record_test_search` for reference.

File: mathesar/api/db/viewsets/records.py (hunk 3 above)
Comment (sahilsuman933) [reply mathesar/api/db/viewsets/records.py]: I have addressed all these changes that you mentioned in my latest commit.

**Comment by sahilsuman933** [comment]
@Anish9901 The test case is failing and it is because the `Patent Expiration Date` is being treated as text rather than DATE as we discussed earlier. So, should I address this particular issue in the same pull request or should I create a new pull request for it.

**Comment by sahilsuman933** [comment]
@Anish9901 Now the tests related to invalid dates are passing. Thank you so much for your guidance :)

**Review by Anish9901 — APPROVED** [review]
Alright, looks good now @sahilsuman933. Thanks for your contribution!

**Pull request merged**